arrows/vxl/pixel_feature_extractor.cxx

```diff
@@ -0,0 +1,321 @@
+
+#include <arrows/vxl/image_container.h>
+#include <vital/config/config_block_io.h>
+
+#include <vil/vil_convert.h>
+#include <vil/vil_image_view.h>
+#include <vil/vil_math.h>
+#include <vil/vil_plane.h>
+
+#include <cstdlib>
+#include <limits>
+#include <type_traits>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+// ----------------------------------------------------------------------------
+// Private implementation class
+class pixel_feature_extractor::priv
+{
+public:
+
+  priv( pixel_feature_extractor* parent ) : p{ parent }
+  {
+  }
+
+  // Copy multiple filtered images into contigious memory
+  template < typename pix_t > vil_image_view< pix_t >
+  concatenate_images( std::vector< vil_image_view< pix_t > > filtered_images );
+  // Extract local pixel-wise features
+  template < typename response_t > vil_image_view< response_t >
+  filter( kwiver::vital::image_container_sptr input_image );
+
+  pixel_feature_extractor* p;
+
+  bool enable_color{ true };
+  bool enable_gray{ true };
+  bool enable_aligned_edge{ true };
+  bool enable_average{ true };
+  bool enable_convert{ true };
+  bool enable_color_commonality{ true };
+  bool enable_high_pass_bidir{ true };
+  bool enable_high_pass_box{ true };
+
+  vxl::aligned_edge_detection aligned_edge_detection_filter;
+  vxl::average_frames average_frames_filter;
+  vxl::convert_image convert_image_filter;
+  vxl::color_commonality_filter color_commonality_filter;
+  vxl::high_pass_filter high_pass_bidir_filter;
+  vxl::high_pass_filter high_pass_box_filter;
+};
+
+// ----------------------------------------------------------------------------
+template < typename pix_t >
+vil_image_view< pix_t >
+pixel_feature_extractor::priv
+::concatenate_images( std::vector< vil_image_view< pix_t > > filtered_images )
+{
+  // Count the total number of planes
+  unsigned total_planes{ 0 };
+
+  for( auto const& image : filtered_images )
+  {
+    total_planes += image.nplanes();
+  }
+
+  if( total_planes == 0 )
+  {
+    LOG_ERROR( p->logger(), "No filtered images provided" );
+    return {};
+  }
+
+  auto const ni = filtered_images.at( 0 ).ni();
+  auto const nj = filtered_images.at( 0 ).nj();
+  vil_image_view< pix_t > concatenated_planes{ ni, nj, total_planes };
+
+  // Concatenate the filtered images into a single output
+  unsigned current_plane{ 0 };
+
+  for( auto const& image : filtered_images )
+  {
+    for( unsigned i{ 0 }; i < image.nplanes(); ++i )
+    {
+      vil_plane( concatenated_planes,
+                 current_plane ).deep_copy( vil_plane( image, i ) );
+      ++current_plane;
+    }
+  }
+  return concatenated_planes;
+}
+
+// ----------------------------------------------------------------------------
+template < typename pix_t >
+vil_image_view< pix_t >
+pixel_feature_extractor::priv
+::filter( kwiver::vital::image_container_sptr input_image )
+{
+  std::vector< vil_image_view< vxl_byte > > filtered_images;
+
+  if( enable_color || enable_gray )
```

Comment:
I actually don't think there is much value in building a sprokit pipeline that runs each of these filters in separate pipeline processes.  These are all independent filters that run in parallel, so it would probably be just as efficient, or more efficient, to make this function multi-threaded (later, not now) to run this processing in parallel.   The only advantage of the sprokit pipeline would be better run-time configurability if you wanted to reorder these or add a new one without recompiling.

Reply:
This makes sense to me.